simulation/optimization/genetic.py

```diff
@@ -257,9 +257,46 @@ def amplifier(race_length):
         # Add a time delay between generations (used for debug purposes)
         delay_after_generation = 0.0
 
-        # Define a function to be run when a generation begins
-        # Here we define it to update the progress bar, if it exists
-        on_generation = (lambda x: pbar.update(1)) if pbar is not None else (lambda x: print("New generation!"))
+        # Store diversity of generation per optimization iteration
+        self.diversity = []
+
+        # Stopping context based on stopping criteria
+        # If stopping criteria = "saturate" -> stopping_point = generations_completed
+        # If stopping criteria = "reach" -> stopping_point = fitness reached
+        self.stopping_point = 0
+
+        # A function to be run when a generation begins
+        def on_generation_callback(x):
+            """
+            Callback function that is called after each generation/optimization iteration.
+            Passes in a GA instance named x in this func
+            """
+
+            # Calculate Diversity
+            # sum accumulator for standard deviation of a stage
+            sum_stage_sd = 0  # stages -> individual gene
+
+            for i in range(x.pop_size[1]):  # iterate through each gene/stage
+                stage_mean = np.mean(x.population[:, i])
+                squared_diffs = np.square(x.population[:, i] - stage_mean)
+                mean_squared_diffs = np.mean(squared_diffs)  # mean of squared differences
+                sum_stage_sd += np.sqrt(mean_squared_diffs)  # add standard deviation of this gene
+
+            # Diversity of this population / generation -> average standard deviation of genes
+            diversity = sum_stage_sd / x.pop_size[1]
+            self.diversity.append(diversity)
+
+            # Record Stopping point info
+            if "saturate" in str(self.settings.stopping_criteria):
```

Comment:
Stopping point should be `x.generations_completed` in both cases. If it stops early because of reaching the target fitness then this fitness will be recorded as the best solution fitness anyway.